README.md
# Upload

---------------

[![spm package](http://spmjs.io/badge/arale-upload)](http://spmjs.io/package/arale-upload)

iframe and html5 uploader.


<span style="font-size:120px;line-height:0.4;color:rgb(255, 140, 42);font-family:Menlo;">⇪</span>

## 演示

查看演示，你需要 clone 一份代码：

```
$ git clone git://github.com/aralejs/upload
$ cd upload
$ npm install
$ node server.js
```

然后访问：http://localhost:8000/demo.html


## Attributes

```javascript
var uploader = new Uploader({
    trigger: '#upload-icon',
    name: 'image',
    action: '/upload',
    accept: 'image/*',
    data: {'xsrf': 'hash'},
    multiple: true,
    useIframe: true,
    error: function(file) {
        alert(file);
    },
    success: function(response) {
        alert(response);
    },
    progress: function(event, position, total, percent, files) {
        console.log(percent);
    }
});
```

### trigger `element|selector`

trigger 唤出文件选择器，可以是：

    - 选择器
    - element
    - jQuery Object

### name `string`

name 即为 `<input name="{{name}}">` 的值，即上传文件时对应的 name。

### action `url`

action 为 `<form action="{{action}}">` 的值，表单提交的地址。

### accept `string`

支持的文件类型，比如 `image/\*` 为只上传图片类的文件。可选项。

### multiple `boolean`

是否支持多文件上传。默认为 false。

### useIframe `boolean`

始终使用Iframe方式上传，默认为false。

### data `object`

随表单一起要提交的数据。

### error `function`

上传失败的回调函数。

### success `function`

上传成功的回调函数。

### progress `function`

上传的进度回调，不支持 IE9-。回调的参数分别为 ajaxEvent, 当前上传字节，总字节，进度百分比和当前文件列表。


## Methods

链式调用:

```javascript
var uploader = new Uploader({
    trigger: '#upload-icon',
    name: 'image',
    action: '/upload',
    data: {'xsrf': 'hash'}
}).success(function(response) {
    alert(response);
}).error(function(file) {
    alert(file);
});
```

## Data API

```html
<a id="upload" data-name="image" data-action="/upload" data-data="a=a&b=b">Upload</a>
<script>
var uploader = new Uploader({'trigger': '#upload'});
// more friendly way
// var uploader = new Uploader('#upload');
uploader.success(function(response) {
    alert(response);
});
</script>
```

## Advanced

Demo in **API** section could not be controlled. When you select a file, it will
be submitted immediately. We can broke the chain with ``change``:

```javascript
var uploader = new Uploader({
    trigger: '#upload-icon',
    name: 'image',
    action: '/upload',
    data: {'xsrf': 'hash'}
}).change(function(files) {
    for (var i=0; i<files.length; i++) {
        console.log('you are selecting ' + files[i].name + ' Size: ' + files[i].size);
    }
    // Default behavior of change is
    // this.submit();
}).success(function(response) {
    alert(response);
});
```

Now you should handle it yourself:

```javascript
$('#submit').click(function() {
    uploader.submit();
});
```


## Show Progress

It is impossible to show progress, but you can make a fake prgress.


```javascript
var uploader = new Uploader({
    trigger: '#upload-icon',
    name: 'image',
    action: '/upload',
    data: {'xsrf': 'hash'},
    progress: function(e, position, total, percent, files) {
      $('#progress').text('Uploading ... ' + percent + '%');
    }
}).change(function(files) {
    // before submit
    $('#progress').text('Uploading ...');
    this.submit();
}).success(function(response) {
    $('#progress').text('Success');
    alert(response);
});
```


## Seajs Hint

Load uploader with seajs:

```javascript
seajs.use('upload', function(Uploader) {
    var uploader = new Uploader({
    });
});
```

## Changelog

**2013-12-10** `1.1.0`

1. Add upload progress for html5 uploader
2. change function argument change to a files list.
3. fix multiple attribute.
4. fix this in change function
5. fix FileList object in ie 9-

**2013-07-18** `1.0.1`

1. Support choosing the same file for uploader
2. Fix memory leaks for FormData

**2013-06-25** `1.0.0`

Combine iframe and html5 uploader.

## IE hint

https://github.com/blueimp/jQuery-File-Upload/issues/1795


HISTORY.md
# History

---

## 1.2.0

`improved` 升级到 spm@3.x 规范。

## 1.1.2

`fixed` 修复跨域 iframe 上传时报错的问题。#37

## 1.1.1

`fixed` #28 修复连续上传多个文件时，进度条显示重复叠加到最新的文件上的问题。

`fixed` #28 修复连续上传多个文件时，在 IE 下 iframe.contents() 调用报错的问题。

`fixed` 修复在 IE 下浏览器上传后仍然处于 loading 状态的[问题](https://github.com/blueimp/jQuery-File-Upload/blob/9.5.6/js/jquery.iframe-transport.js#L102)

## 1.1.0

`improved` [添加进度](https://github.com/aralejs/upload/issues/23)

`improved` [添加 enable/disable](https://github.com/aralejs/upload/pull/25)

## 1.0.1



## 1.0.0

`new` 初始化


server.js
var fs = require('fs');
var http = require('http');

var server = http.createServer(function(req, res) {
  console.log(req.method + ' ' + req.url);

  if (req.method === 'POST') {
    var data = {
      length: req.headers['content-length']
    };

    if (req.headers['x-requested-with'] === 'XMLHttpRequest') {
      data['uploader'] = 'html5';
    } else {
      data['uploader'] = 'iframe';
    }

    req.on('data', function() {
    });

    console.log(data);

    req.on('end', function() {

      res.writeHead(200);

      res.end(JSON.stringify(data));

    });

  } else {

    req.addListener('end', function() {
      var Server = require('node-static').Server;
      var file = new Server(fs.realpathSync('.'));

      file.serve(req, res);
    }).resume();
  }


});
console.log('View demo at http://localhost:8000/demo.html');
server.listen(8000);


server.py
#!/usr/bin/env python

import tornado.httpserver
import tornado.ioloop
import tornado.options
import tornado.web

from tornado.options import define, options

define("port", default=8888, help="run on the given port", type=int)


class MainHandler(tornado.web.RequestHandler):
    def get(self):
        f = open('assets/demo.html')
        content = f.read()
        f.close()
        self.write(content)

    def post(self):
        print(self.request.arguments)
        f = self.request.files.get('file', None)
        if not f:
            self.write('{"stat": "fail", "msg": "no file"}')
            return
        f = f[0]
        filename = f.get('filename', '')
        self.write('{"stat": "ok", "msg": "get file %s"}' % filename)


def main():
    tornado.options.parse_command_line()
    application = tornado.web.Application([
        (r"/", MainHandler),
        ("/src/(.*)", tornado.web.StaticFileHandler, {'path': 'src/'}),
        ("/assets/(.*)", tornado.web.StaticFileHandler, {'path': 'assets/'}),
    ])
    http_server = tornado.httpserver.HTTPServer(application)
    http_server.listen(options.port)
    tornado.ioloop.IOLoop.instance().start()


if __name__ == "__main__":
    print("Starting server at: 127.0.0.1:%s" % options.port)
    main()


src/upload.js
define(function(require, exports, module) {
  var $ = require('$');
  var iframeCount = 0;

  function Uploader(options) {
    if (!(this instanceof Uploader)) {
      return new Uploader(options);
    }
    if (isString(options)) {
      options = {trigger: options};
    }

    var settings = {
      trigger: null,
      name: null,
      action: null,
      data: null,
      accept: null,
      change: null,
      error: null,
      multiple: true,
      success: null
    };
    if (options) {
      $.extend(settings, options);
    }
    var $trigger = $(settings.trigger);

    settings.action = settings.action || $trigger.data('action') || '/upload';
    settings.name = settings.name || $trigger.attr('name') || $trigger.data('name') || 'file';
    settings.data = settings.data || parse($trigger.data('data'));
    settings.accept = settings.accept || $trigger.data('accept');
    settings.success = settings.success || $trigger.data('success');
    this.settings = settings;

    this.setup();
    this.bind();
  }

  // initialize
  // create input, form, iframe
  Uploader.prototype.setup = function() {
    this.form = $(
      '<form method="post" enctype="multipart/form-data"'
      + 'target="" action="' + this.settings.action + '" />'
    );

    this.iframe = newIframe();
    this.form.attr('target', this.iframe.attr('name'));

    var data = this.settings.data;
    this.form.append(createInputs(data));
    if (window.FormData) {
      this.form.append(createInputs({'_uploader_': 'formdata'}));
    } else {
      this.form.append(createInputs({'_uploader_': 'iframe'}));
    }

    var input = document.createElement('input');
    input.type = 'file';
    input.name = this.settings.name;
    if (this.settings.accept) {
      input.accept = this.settings.accept;
    }
    if (this.settings.multiple) {
      input.multiple = true;
      input.setAttribute('multiple', 'multiple');
    }
    this.input = $(input);

    var $trigger = $(this.settings.trigger);
    this.input.attr('hidefocus', true).css({
      position: 'absolute',
      top: 0,
      right: 0,
      opacity: 0,
      outline: 0,
      cursor: 'pointer',
      height: $trigger.outerHeight(),
      fontSize: Math.max(64, $trigger.outerHeight() * 5)
    });
    this.form.append(this.input);
    this.form.css({
      position: 'absolute',
      top: $trigger.offset().top,
      left: $trigger.offset().left,
      overflow: 'hidden',
      width: $trigger.outerWidth(),
      height: $trigger.outerHeight(),
      zIndex: findzIndex($trigger) + 10
    }).appendTo('body');
    return this;
  };

  // bind events
  Uploader.prototype.bind = function() {
    var self = this;
    var $trigger = $(self.settings.trigger);
    $trigger.mouseenter(function() {
      self.form.css({
        top: $trigger.offset().top,
        left: $trigger.offset().left,
        width: $trigger.outerWidth(),
        height: $trigger.outerHeight()
      });
    });
    self.bindInput();
  };

  Uploader.prototype.bindInput = function() {
    var self = this;
    self.input.change(function(e) {
      // ie9 don't support FileList Object
      // http://stackoverflow.com/questions/12830058/ie8-input-type-file-get-files
      self._files = this.files || [{
        name: e.target.value
      }];
      var file = self.input.val();
      if (self.settings.change) {
        self.settings.change.call(self, self._files);
      } else if (file) {
        return self.submit();
      }
    });
  };

  // handle submit event
  // prepare for submiting form
  Uploader.prototype.submit = function() {
    var self = this;
    if (window.FormData && self._files) {
      // build a FormData
      var form = new FormData(self.form.get(0));
      // use FormData to upload
      form.append(self.settings.name, self._files);

      var optionXhr;
      if (self.settings.progress) {
        // fix the progress target file
        var files = self._files;
        optionXhr = function() {
          var xhr = $.ajaxSettings.xhr();
          if (xhr.upload) {
            xhr.upload.addEventListener('progress', function(event) {
              var percent = 0;
              var position = event.loaded || event.position; /*event.position is deprecated*/
              var total = event.total;
              if (event.lengthComputable) {
                  percent = Math.ceil(position / total * 100);
              }
              self.settings.progress(event, position, total, percent, files);
            }, false);
          }
          return xhr;
        };
      }
      $.ajax({
        url: self.settings.action,
        type: 'post',
        processData: false,
        contentType: false,
        data: form,
        xhr: optionXhr,
        context: this,
        success: self.settings.success,
        error: self.settings.error
      });
      return this;
    } else {
      // iframe upload
      self.iframe = newIframe();
      self.form.attr('target', self.iframe.attr('name'));
      $('body').append(self.iframe);
      self.iframe.one('load', function() {
        // https://github.com/blueimp/jQuery-File-Upload/blob/9.5.6/js/jquery.iframe-transport.js#L102
        // Fix for IE endless progress bar activity bug
        // (happens on form submits to iframe targets):
        $('<iframe src="javascript:false;"></iframe>')
          .appendTo(self.form)
          .remove();
        var response;
        try {
          response = $(this).contents().find('body').html();
        } catch (e) {
          response = 'cross-domain';
        }
        $(this).remove();
        if (!response) {
          if (self.settings.error) {
            self.settings.error(self.input.val());
          }
        } else {
          if (self.settings.success) {
            self.settings.success(response);
          }
        }
      });
      self.form.submit();
    }
    return this;
  };

  Uploader.prototype.refreshInput = function() {
    //replace the input element, or the same file can not to be uploaded
    var newInput = this.input.clone();
    this.input.before(newInput);
    this.input.off('change');
    this.input.remove();
    this.input = newInput;
    this.bindInput();
  };

  // handle change event
  // when value in file input changed
  Uploader.prototype.change = function(callback) {
    if (!callback) {
      return this;
    }
    this.settings.change = callback;
    return this;
  };

  // handle when upload success
  Uploader.prototype.success = function(callback) {
    var me = this;
    this.settings.success = function(response) {
      me.refreshInput();
      if (callback) {
        callback(response);
      }
    };

    return this;
  };

  // handle when upload success
  Uploader.prototype.error = function(callback) {
    var me = this;
    this.settings.error = function(response) {
      if (callback) {
        me.refreshInput();
        callback(response);
      }
    };
    return this;
  };

  // enable
  Uploader.prototype.enable = function(){
    this.input.prop('disabled', false);
  };

  // disable
  Uploader.prototype.disable = function(){
    this.input.prop('disabled', true);
  };

  // Helpers
  // -------------

  function isString(val) {
    return Object.prototype.toString.call(val) === '[object String]';
  }

  function createInputs(data) {
    if (!data) return [];

    var inputs = [], i;
    for (var name in data) {
      i = document.createElement('input');
      i.type = 'hidden';
      i.name = name;
      i.value = data[name];
      inputs.push(i);
    }
    return inputs;
  }

  function parse(str) {
    if (!str) return {};
    var ret = {};

    var pairs = str.split('&');
    var unescape = function(s) {
      return decodeURIComponent(s.replace(/\+/g, ' '));
    };

    for (var i = 0; i < pairs.length; i++) {
      var pair = pairs[i].split('=');
      var key = unescape(pair[0]);
      var val = unescape(pair[1]);
      ret[key] = val;
    }

    return ret;
  }

  function findzIndex($node) {
    var parents = $node.parentsUntil('body');
    var zIndex = 0;
    for (var i = 0; i < parents.length; i++) {
      var item = parents.eq(i);
      if (item.css('position') !== 'static') {
        zIndex = parseInt(item.css('zIndex'), 10) || zIndex;
      }
    }
    return zIndex;
  }

  function newIframe() {
    var iframeName = 'iframe-uploader-' + iframeCount;
    var iframe = $('<iframe name="' + iframeName + '" />').hide();
    iframeCount += 1;
    return iframe;
  }

  function MultipleUploader(options) {
    if (!(this instanceof MultipleUploader)) {
      return new MultipleUploader(options);
    }

    if (isString(options)) {
      options = {trigger: options};
    }
    var $trigger = $(options.trigger);

    var uploaders = [];
    $trigger.each(function(i, item) {
      options.trigger = item;
      uploaders.push(new Uploader(options));
    });
    this._uploaders = uploaders;
  }
  MultipleUploader.prototype.submit = function() {
    $.each(this._uploaders, function(i, item) {
      item.submit();
    });
    return this;
  };
  MultipleUploader.prototype.change = function(callback) {
    $.each(this._uploaders, function(i, item) {
      item.change(callback);
    });
    return this;
  };
  MultipleUploader.prototype.success = function(callback) {
    $.each(this._uploaders, function(i, item) {
      item.success(callback);
    });
    return this;
  };
  MultipleUploader.prototype.error = function(callback) {
    $.each(this._uploaders, function(i, item) {
      item.error(callback);
    });
    return this;
  };
  MultipleUploader.prototype.enable = function (){
    $.each(this._uploaders, function (i, item){
      item.enable();
    });
    return this;
  };
  MultipleUploader.prototype.disable = function (){
    $.each(this._uploaders, function (i, item){
      item.disable();
    });
    return this;
  };
  MultipleUploader.Uploader = Uploader;

  module.exports = MultipleUploader;
});


src/iframe-uploader.js
// iframe uploader
// ===============
//

define(function(require, exports, module) {
    var $ = require('jquery');

    function IframeUploader(options) {
        if (!(this instanceof IframeUploader)) {
            return new IframeUploader(options);
        }
        if (isString(options)) options = {trigger: options};

        var settings = {
            trigger: null,
            name: null,
            action: null,
            data: null,
            accept: null,
            change: null,
            error: null,
            success: null
        };
        if (options) {
            $.extend(settings, options);
        }
        var $trigger = $(settings.trigger);

        settings.action = settings.action || $trigger.data('action') ||
            '/upload';
        settings.name = settings.name || $trigger.data('name') || 'file';
        settings.data = settings.data || parse($trigger.data('data'));
        settings.accept = settings.accept || $trigger.data('accept');
        settings.success = settings.success || $trigger.data('success');
        this.settings = settings;

        this.setup();
        this.bind();
    }

    // initialize
    // create input, form, iframe
    IframeUploader.prototype.setup = function() {
        var iframeName = 'iframe-uploader-' + new Date().getTime();
        this.iframe = $('<iframe name="' + iframeName + '" />').hide();

        this.form = $('<form method="post" enctype="multipart/form-data"'
                      + 'target="' + iframeName + '" '
                      + 'action="' + this.settings.action + '" />');

        this.form.append(createInputs(this.settings.data));

        var input = document.createElement('input');
        input.type = 'file'; input.name = this.settings.name;
        if (this.settings.accept) input.accept = this.settings.accept;
        this.input = $(input);

        var $trigger = $(this.settings.trigger);
        this.input.attr('hidefocus', true).css({
            position: 'absolute',
            top: 0,
            right: 0,
            opacity: 0,
            outline: 0,
            cursor: 'pointer',
            height: $trigger.outerHeight(),
            fontSize: Math.max(64, $trigger.outerHeight() * 5)
        });
        this.form.append(this.input);
        this.form.css({
            position: 'absolute',
            top: $trigger.offset().top,
            left: $trigger.offset().left,
            overflow: 'hidden',
            width: $trigger.outerWidth(),
            height: $trigger.outerHeight(),
            zIndex: findzIndex($trigger) + 10
        }).appendTo('body');

        return this;
    };

    // bind events
    IframeUploader.prototype.bind = function() {
        var self = this;
        var $trigger = $(self.settings.trigger);
        $trigger.mouseenter(function() {
            self.form.css({
              top: $trigger.offset().top,
              left: $trigger.offset().left,
              width: $trigger.outerWidth(),
              height: $trigger.outerHeight()
            })
        });
        self.input.change(function() {
            var file = self.input.val();
            if (self.settings.change) {
              if (file) file = file.substr(file.lastIndexOf('\\') + 1);
              self.settings.change(file);
            } else if (file) {
              return self.submit();
            }
        });
    };

    // handle submit event
    // prepare for submiting form
    IframeUploader.prototype.submit = function() {
        var self = this;
        $('body').append(self.iframe);
        self.iframe.on('load', function() {
            var response = self.iframe.contents().find('body').html();
            self.iframe.off('load').remove();
            if (!response) {
                if (self.settings.error) self.settings.error(self.input.val());
            } else {
                if (self.settings.success) self.settings.success(response);
            }
        });
        self.form.submit();
        return this;
    };

    // handle change event
    // when value in file input changed
    IframeUploader.prototype.change = function(callback) {
        if (!callback) return this;
        this.settings.change = callback;
        return this;
    };

    // handle when upload success
    IframeUploader.prototype.success = function(callback) {
        if (!callback) return this;
        this.settings.success = callback;
        return this;
    };

    // handle when upload success
    IframeUploader.prototype.error = function(callback) {
        if (!callback) return this;
        this.settings.error = callback;
        return this;
    };

    // Helpers
    // -------------

    function isString(val) {
        return Object.prototype.toString.call(val) === '[object String]';
    }

    function createInputs(data) {
        if (!data) return [];

        var inputs = [], i;
        for (var name in data) {
            i = document.createElement('input');
            i.type = 'hidden';
            i.name = name;
            i.value = data[name];
            inputs.push(i);
        }
        return inputs;
    }

    function parse(str) {
        if (!str) return {};
        var ret = {};

        var pairs = str.split('&');
        var unescape = function(s) {
            return decodeURIComponent(s.replace(/\+/g, ' '));
        };

        for (var i = 0; i < pairs.length; i++) {
            var pair = pairs[i].split('=');
            var key = unescape(pair[0]);
            var val = unescape(pair[1]);
            ret[key] = val;
        }

        return ret;
    }

    function findzIndex($node) {
        var parents = $node.parentsUntil('body');
        var zIndex = 0;
        for (var i = 0; i < parents.length; i++) {
            var item = parents.eq(i);
            if (item.css('position') !== 'static') {
                zIndex = parseInt(item.css('zIndex'), 10) || zIndex;
            }
        }
        return zIndex;
    }

    module.exports = IframeUploader;
});


index.js
var $ = require('jquery');
var iframeCount = 0;

function Uploader(options) {
  if (!(this instanceof Uploader)) {
    return new Uploader(options);
  }
  if (isString(options)) {
    options = {trigger: options};
  }

  var settings = {
    trigger: null,
    name: null,
    action: null,
    data: null,
    accept: null,
    change: null,
    error: null,
    multiple: true,
    success: null
  };
  if (options) {
    $.extend(settings, options);
  }
  var $trigger = $(settings.trigger);

  settings.action = settings.action || $trigger.data('action') || '/upload';
  settings.name = settings.name || $trigger.attr('name') || $trigger.data('name') || 'file';
  settings.data = settings.data || parse($trigger.data('data'));
  settings.accept = settings.accept || $trigger.data('accept');
  settings.success = settings.success || $trigger.data('success');
  this.settings = settings;

  this.setup();
  this.bind();
}

// initialize
// create input, form, iframe
Uploader.prototype.setup = function() {
  this.form = $(
    '<form method="post" enctype="multipart/form-data"'
    + 'target="" action="' + this.settings.action + '" />'
  );

  this.iframe = newIframe();
  this.form.attr('target', this.iframe.attr('name'));

  var data = this.settings.data;
  this.form.append(createInputs(data));
  if (window.FormData) {
    this.form.append(createInputs({'_uploader_': 'formdata'}));
  } else {
    this.form.append(createInputs({'_uploader_': 'iframe'}));
  }

  var input = document.createElement('input');
  input.type = 'file';
  input.name = this.settings.name;
  if (this.settings.accept) {
    input.accept = this.settings.accept;
  }
  if (this.settings.multiple) {
    input.multiple = true;
    input.setAttribute('multiple', 'multiple');
  }
  this.input = $(input);

  var $trigger = $(this.settings.trigger);
  this.input.attr('hidefocus', true).css({
    position: 'absolute',
    top: 0,
    right: 0,
    opacity: 0,
    outline: 0,
    cursor: 'pointer',
    height: $trigger.outerHeight(),
    fontSize: Math.max(64, $trigger.outerHeight() * 5)
  });
  this.form.append(this.input);
  this.form.css({
    position: 'absolute',
    top: $trigger.offset().top,
    left: $trigger.offset().left,
    overflow: 'hidden',
    width: $trigger.outerWidth(),
    height: $trigger.outerHeight(),
    zIndex: findzIndex($trigger) + 10
  }).appendTo('body');
  return this;
};

// bind events
Uploader.prototype.bind = function() {
  var self = this;
  var $trigger = $(self.settings.trigger);
  $trigger.mouseenter(function() {
    self.form.css({
      top: $trigger.offset().top,
      left: $trigger.offset().left,
      width: $trigger.outerWidth(),
      height: $trigger.outerHeight()
    });
  });
  self.bindInput();
};

Uploader.prototype.bindInput = function() {
  var self = this;
  self.input.change(function(e) {
    // ie9 don't support FileList Object
    // http://stackoverflow.com/questions/12830058/ie8-input-type-file-get-files
    self._files = this.files || [{
      name: e.target.value
    }];
    var file = self.input.val();
    if (self.settings.change) {
      self.settings.change.call(self, self._files);
    } else if (file) {
      return self.submit();
    }
  });
};

// handle submit event
// prepare for submiting form
Uploader.prototype.submit = function() {
  var self = this;
  if (window.FormData && self._files) {
    // build a FormData
    var form = new FormData(self.form.get(0));
    // use FormData to upload
    form.append(self.settings.name, self._files);

    var optionXhr;
    if (self.settings.progress) {
      // fix the progress target file
      var files = self._files;
      optionXhr = function() {
        var xhr = $.ajaxSettings.xhr();
        if (xhr.upload) {
          xhr.upload.addEventListener('progress', function(event) {
            var percent = 0;
            var position = event.loaded || event.position; /*event.position is deprecated*/
            var total = event.total;
            if (event.lengthComputable) {
                percent = Math.ceil(position / total * 100);
            }
            self.settings.progress(event, position, total, percent, files);
          }, false);
        }
        return xhr;
      };
    }
    $.ajax({
      url: self.settings.action,
      type: 'post',
      processData: false,
      contentType: false,
      data: form,
      xhr: optionXhr,
      context: this,
      success: self.settings.success,
      error: self.settings.error
    });
    return this;
  } else {
    // iframe upload
    self.iframe = newIframe();
    self.form.attr('target', self.iframe.attr('name'));
    $('body').append(self.iframe);
    self.iframe.one('load', function() {
      // https://github.com/blueimp/jQuery-File-Upload/blob/9.5.6/js/jquery.iframe-transport.js#L102
      // Fix for IE endless progress bar activity bug
      // (happens on form submits to iframe targets):
      $('<iframe src="javascript:false;"></iframe>')
        .appendTo(self.form)
        .remove();
      var response;
      try {
        response = $(this).contents().find("body").html();
      } catch (e) {
        response = "cross-domain";
      }
      $(this).remove();
      if (!response) {
        if (self.settings.error) {
          self.settings.error(self.input.val());
        }
      } else {
        if (self.settings.success) {
          self.settings.success(response);
        }
      }
    });
    self.form.submit();
  }
  return this;
};

Uploader.prototype.refreshInput = function() {
  //replace the input element, or the same file can not to be uploaded
  var newInput = this.input.clone();
  this.input.before(newInput);
  this.input.off('change');
  this.input.remove();
  this.input = newInput;
  this.bindInput();
};

// handle change event
// when value in file input changed
Uploader.prototype.change = function(callback) {
  if (!callback) {
    return this;
  }
  this.settings.change = callback;
  return this;
};

// handle when upload success
Uploader.prototype.success = function(callback) {
  var me = this;
  this.settings.success = function(response) {
    me.refreshInput();
    if (callback) {
      callback(response);
    }
  };

  return this;
};

// handle when upload success
Uploader.prototype.error = function(callback) {
  var me = this;
  this.settings.error = function(response) {
    if (callback) {
      me.refreshInput();
      callback(response);
    }
  };
  return this;
};

// enable
Uploader.prototype.enable = function(){
  this.input.prop('disabled', false);
};

// disable
Uploader.prototype.disable = function(){
  this.input.prop('disabled', true);
};

// Helpers
// -------------

function isString(val) {
  return Object.prototype.toString.call(val) === '[object String]';
}

function createInputs(data) {
  if (!data) return [];

  var inputs = [], i;
  for (var name in data) {
    i = document.createElement('input');
    i.type = 'hidden';
    i.name = name;
    i.value = data[name];
    inputs.push(i);
  }
  return inputs;
}

function parse(str) {
  if (!str) return {};
  var ret = {};

  var pairs = str.split('&');
  var unescape = function(s) {
    return decodeURIComponent(s.replace(/\+/g, ' '));
  };

  for (var i = 0; i < pairs.length; i++) {
    var pair = pairs[i].split('=');
    var key = unescape(pair[0]);
    var val = unescape(pair[1]);
    ret[key] = val;
  }

  return ret;
}

function findzIndex($node) {
  var parents = $node.parentsUntil('body');
  var zIndex = 0;
  for (var i = 0; i < parents.length; i++) {
    var item = parents.eq(i);
    if (item.css('position') !== 'static') {
      zIndex = parseInt(item.css('zIndex'), 10) || zIndex;
    }
  }
  return zIndex;
}

function newIframe() {
  var iframeName = 'iframe-uploader-' + iframeCount;
  var iframe = $('<iframe name="' + iframeName + '" />').hide();
  iframeCount += 1;
  return iframe;
}

function MultipleUploader(options) {
  if (!(this instanceof MultipleUploader)) {
    return new MultipleUploader(options);
  }

  if (isString(options)) {
    options = {trigger: options};
  }
  var $trigger = $(options.trigger);

  var uploaders = [];
  $trigger.each(function(i, item) {
    options.trigger = item;
    uploaders.push(new Uploader(options));
  });
  this._uploaders = uploaders;
}
MultipleUploader.prototype.submit = function() {
  $.each(this._uploaders, function(i, item) {
    item.submit();
  });
  return this;
};
MultipleUploader.prototype.change = function(callback) {
  $.each(this._uploaders, function(i, item) {
    item.change(callback);
  });
  return this;
};
MultipleUploader.prototype.success = function(callback) {
  $.each(this._uploaders, function(i, item) {
    item.success(callback);
  });
  return this;
};
MultipleUploader.prototype.error = function(callback) {
  $.each(this._uploaders, function(i, item) {
    item.error(callback);
  });
  return this;
};
MultipleUploader.prototype.enable = function (){
  $.each(this._uploaders, function (i, item){
    item.enable();
  });
  return this;
};
MultipleUploader.prototype.disable = function (){
  $.each(this._uploaders, function (i, item){
    item.disable();
  });
  return this;
};
MultipleUploader.Uploader = Uploader;

module.exports = MultipleUploader;
